File: College Study/sem2/DS/Merge Sort/MergeSort_Array.cpp
#include <iostream>
#include <stdio.h>

using namespace std;
int* merge(int* Arr ,int start , int mid , int end) {
    int mergedArray[end-start+1];
    int i,j,k;
    for(i=start,j=mid+1,k=0;i<=mid && j<=end;) {
        if(Arr[i]<Arr[j]) {
            mergedArray[k] = Arr[i];
            ++i;
        }
        else {
            mergedArray[k] = Arr[j];
            ++j;
        }
        ++k;
    }

    while(i<=mid) {
        mergedArray[k++] = Arr[i++];
    }
    while(j<=end) {
        mergedArray[k++] = Arr[j++];
    }

    for(k=0;k<end-start+1;++k) {
        Arr[start+k] = mergedArray[k];
    }

}

void mergeSort(int* Arr , int start , int end) {

    if(end-start <=0) return;

    int mid = (start+end)/2;

    mergeSort(Arr,start,mid);
    mergeSort(Arr,mid+1,end);

    merge(Arr,start,mid,end);

}

int main()
{
    int Arr[1000];

    cout << "\nEnter no of Elements :";
    int N;
    cin >> N;
    cout << "Enter Elements";
    for (int i = 0; i < N; ++i)
    {
        scanf("%d", &Arr[i]);
    } 

    mergeSort(Arr, 0 , N-1);

    cout << "\nSorted Array :..\n";

    for (int i = 0; i < N; ++i)
    {
        printf(" %d", Arr[i]);
    }
    return 0;
}

File: College Study/sem2/DS/Merge Sort/mergeSort_Array.py
#!/usr/bin/python3

def merge(Arr1 , Arr2) :
    mergedList = list()
    i=j=0
    while(i<len(Arr1) and j<len(Arr2)) :
        if(Arr1[i]<Arr2[j]) :
            mergedList.append(Arr1[i])
            i+=1
        else :
            mergedList.append(Arr2[j])
            j+=1

    while i<len(Arr1) :
        mergedList.append(Arr1[i])
        i+=1
    while j<len(Arr2) :
        mergedList.append(Arr2[j])
        j+=1

    return mergedList    

def mergeSort(Arr) :
    if(len(Arr)<=1) : 
        return Arr
    if(len(Arr)==2) :
        return merge(Arr[0:1] , Arr[1:2])

    dividePivot = len(Arr)//2

    Arr = merge(mergeSort(Arr[:dividePivot]) , mergeSort(Arr[dividePivot:]))

    return Arr


if __name__ == "__main__" :
    print("\n Enter elements for Array 1 (space seperated) \n ")
    Arr1 = [int(i) for i in input().split()]
    # print("\n Enter elements for Array 2 (space seperated) \n ")
    # Arr2 = [int(i) for i in input().split()]
    # print(merge(Arr1,Arr2))
    Arr1 = mergeSort(Arr1)
    print(Arr1)
